### DISCUSSION TITLE
Rate limiting is being introduced to the Octopus APIs

### DISCUSSION THREAD

**Discussion by adds68** [BottlecapDave/HomeAssistant-OctopusEnergy#255] [General]
Hi,

I'm just starting a discussion here to alert the community of some new rate limits being introduced on the Octopus APIs.

> A consumer can make no more than 100 calls per hour to the telemetry API

This does include the mobile app usage also.

**Comment by BottlecapDave** [comment]
Thanks for raising this. Which parts count as the telemetry API? Is this for the graphql part that supports the Home Mini or something else?

**Reply by BottlecapDave** [reply]
@adds68 Is there any update on this?

**Reply by adds68** [reply]
hi @BottlecapDave , yes sorry, it's the whole graphql part 👍

**Reply by BottlecapDave** [reply]
So is this for a single query (no matter how many fields) or only when `smartMeterTelemetry` has been requested? I assume this integration can stay as is out of the box as I'm only requesting data once a minute?

**Reply by dragon2611** [reply]
@BottlecapDave Is it possible to make it configurable to be less frequent, I also use other apps (such as octo-aid) and I think I might be hitting the limit as sometimes that can't get data.

**Reply by adds68** [reply]
Sadly this will not be the entire API, so in theory you can request it minutely with the 100 credits, but as @dragon2611 this will then use up those credits for other apps, including the official app.

**Reply by BottlecapDave** [reply]
Ok I'll add something to the backlog so you can refine how oftern mini data is requested.

> Sadly this will not be the entire AP

Why sadly?

**Reply by Jilas** [reply]
Sorry to wake up this thread.  Just been hit with this problem I think.

Been having oddities with the app for a while (maybe this long in fact) and tonight noticed my HA had at most 25 minutes of data per hour.!!  (Never any gas)

And tonight verified that the app fails at the same time.

Ok so yes, I was testing two versions of the addon (reasons for another post maybe bug tomorrow), but either way that would then just give 50 minutes per hour.

I have this as an active call with them, and the guy in dealing with seems to work closely with the mini team (or in it), so have asked the obvious questions (some I think we mentioned earlier). But especially why change it so low, and why is the app even included within the limit.  Most importantly can they raise it, as seems too low.  If it's 100 calls per hour, as mentioned, that looks to be 50 minutes (that fits observed results) assuming a call every  30 seconds for an update.   But of course it's every minute here so... Something is strange somewhere.

I'll post again when I get some answers.  Just wanted to share what I found today.

Edited (following day).

So I tried running just a single HA instance with the addon, and as predicted above, it lasted 51 minutes before it lost connection.  I've forward this information onto Octopus, as this is clearly not working.  The data is 10 seconds from the mini, and pulling that 6 times slower should be fine.   They must have having some demand issues, and need to scale up the servers to handle the load.  Very disappointed with Octopus for doing this, and while it was noted here by someone that it was happening, I can't see anything about it and I see this as a breaking change and should be documented in their API.  Especially as it breaks their app as well.

**Reply by BottlecapDave** [reply]
Indeed it is a shame. I'm working on a fix, but I'm trying to bundle up a bunch of breaking changes to minimise disruption in the integration. I'll be shifting the updates to once every 2 minutes by default (due to electricity and gas being a possibility through the mini), but making it configurable if the user needs to (as suggested by @adds68. I'm sorry that this wasn't in the integration before this started becoming an issue.

**Reply by Jilas** [reply]
> Indeed it is a shame. I'm working on a fix, but I'm trying to bundle up a bunch of breaking changes to minimise disruption in the integration. I'll be shifting the updates to once every 2 minutes by default (due to electricity and gas being a possibility through the mini), but making it configurable if the user needs to (as suggested by @adds68. I'm sorry that this wasn't in the integration before this started becoming an issue.

I'll keep pushing Octopus now I have someone that gets the problem, was asking me the right questions and listening.  

I think the idea of having it configurable would be a great solution.. Just in case they do something in the future.  

It would be interesting to see what data you get doing it this way,  does it do a summary of all the 10sec data points between pulls.

Just when it fails... You don't get any missing data back.  Gas at this time of year is low use for us and while I know we've used it, nothing came in at all because I missed the one update due to them blocking the connection. 


And certainly don't apologise for something outside of your control.  Until they made a change it was fine and pulling data 6 times slower than the mini collects is very acceptable use of the api in my mind.   (Maybe if you can give us the ability to adjust it, if we have problems we can adjust up, but leave it at 1 minute default, especially for the electric only users).

You've created a great addon.  Hopefully we can get them to change their minds.

**Reply by BottlecapDave** [reply]
The frequency of the Home Mini updates can now be configured when setting up (or updating) your account within the integration in the latest release https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/releases/tag/v8.0.0

**Reply by Jilas** [reply]
Nice work on adding that and all of that other changes.

Still chasing Octopus as I don't see 1 minute requests for data should use up all of those 100 connections in 50 minutes.

The whole point of the mini is to let us get this data so why then limit us.

Their last response took two weeks, when I confirmed all of my testing, (this after they were confused as to even why I had an issue in the first place).  Finally came back with was choice message


> So sorry for the delay - I don't know where time is going but the tech team have come back to me. 
> 
> You are right, we do have rate limiting in place to protect our platforms. We haven't implemented local access yet but never say never!


I did ask about local access while waiting to hear back.

So I've asked them how we can get info back to say this cap is too low and request an increase to make their mini useful...  That was a week ago, so I'll chase on Monday again.

I do understand the need to protect their services, although Thursday (or was it Wednesday) which looks like things failed in that area totally with the outages between 9.30-10.30.  but if they are rolling this out then they need to increase their server capacity (in both directions) but maybe local data will spin out from this.

More when I hear back,

**Reply by Jilas** [reply]
I've had some feedback, and need to send some information over later today, but certainly talking to the right people now.  So fingers crossed.

**Comment by andyblac** [comment]
any more update on this, do you have any direct contacts, maybe few of us can voice our opinion too.

**Reply by Jilas** [reply]
Sorry was hoping for a quick update but seems their change control is much slower than I guessed (which I guess is a good thing).

So I was contacted by James Eddison their CTO, after I think their support teams felt they couldn't add to the conversation they were having with me.

He confirmed they made an arbitrary choice with the cap based on the overall API usage.  So for us that limit was a tad too low.  From my figures it's about 10 minutes to low, so an increase of 20-25% would be a good starting point.  And he had indicated that raising this cap by 25% would be possible, and it had been put in as a change request.

That was on the 17th August.    I have not seen any change to date,  and not knowing their change control cycle, it's possible it's still on the list and just not actioned yet.   I'll drop him another email next week and see if there is an update.  Maybe a combo of holidays and buying Shell customers has delayed things.

**Comment by andyblac** [comment]
@BottlecapDave does the integration make multiple API calls (one for each Electric / GAS) info?, if so ATM we have a setting "Home Mini refresh rate in minuets" could this be split into 2, 1 for each API call ? this means we could keep it `1` for Electric, but allow to be set to `5` or `10` for GAS, as gas does not really need it to be set to 1 as as it does not support live readings. This would reduce the number of API calls.

was just a thought.

**Reply by Jilas** [reply]
Been pondering this, as it#s an interesting idea and could be a potential solution - I've still not heard back from Octopus about when the change request will go in, that should help, but its a big company, and this is a low priority change - so I don't expect it to happen quickly.

Some info to help, currently my gas usage appears 5 minutes past the hour & half hour - although looking back a month ago, it was a different spot in the hour  23/52 past the hour but not sure when I was running out of data then.  Currently I hit the cap now between 10 minutes past and 21 minutes past each hour.

So the idea of having two poll cycles, could be a good solution, as we could keep 1 minute for electric, and having the option to adjust gas independently can then drop that to 5 or 10 minutes as @andyblac suggests (or maybe 15 minutes to be really daring).

@BottlecapDave - I'd certainly be up to help test this, if/when you have time available.  And if there is some basic tests, pre-coding checks or tests, you need doing first, please ask, as would be really happy to help.

Just would be nice to get a resolution to this before winter in the meantime, I'll keep chasing Octopus (their email addresses are pretty easy to work out if anyone wants to send in as well).

**Comment by dragon2611** [comment]
I don't believe the GAS meters send data any more frequently than once every 30 minutes, so I'm not sure why we'd need to poll every 5 minutes or so?

**Reply by andyblac** [reply]
trouble is as far as I understand it, there no fixed time the update happens, so if we would only poll once every 30mins we could miss the update for that 30mins block.

**Reply by Jilas** [reply]
@andyblac @dragon2611 you are both correct.  Gas meters only send out every 30 minutes ('ish).  The question I just finished typing up to Octopus about as well, is the value the api gets for gas is it kept there until its next updated or polled, so you might only need to poll once every 30 minutes.

So with the electric data, I think when you poll it you get the most current version. So if you have electric set to poll every 5 minutes, and boil a kettle in the gap between polls, that usage isn't seen by the api, only the last one.

Assuming the kettle example is right, then polling gas every 30 minutes, should give the last updated value.

Hence asking that question of octopus, and offering to help test the idea with Dave if needed.

I've had gas at different times in the hour but it's also been regular...what made it change (most likely a combo of adjust timings, not running two HA instances with the same Octopus API in real time) but I've not gone over a months worth of data, and comparing with reboots of two HA etc to confirm.  But currently I get them 5/35 minutes part and I lose data between 10-21 minutes each hour.

**Comment by BottlecapDave** [comment]
The API supports gathering data in 10 second increments. However the integration requests the data from the API in 30 minute increments to reduce the volume of data that will be retrieved. If newer data is available, it will pull this information in 30 minute incremenets. This is done to match the rate information and prevent instability with the integration. This will not be changed.

If the polling of gas was changed to every 30 minutes, there is a chance the integration will not pull the data for up to 29 minutes (as data could update 1 minute after the last poll). This can have knock on effects to the energy dashboard if the sensor is used as the data could then be assigned to the wrong hour (due to how the energy dashboard works). 

I'll take a look at separating the polling configuration between electricity and gas as suggested by @andyblac and updating the docs/faq with suggestions. Because the current implementation fixes the rate limiting issue (you just need to decrease the frequency from 1 to 2 minutes), just not perhaps updating data as frequently as one would like, this will probably be behind other issues that are more urgent or effect more people.

**Reply by Jilas** [reply]
Thanks for the detailed response.   Totally understand there are other priorities as well.  The offer of help stands of you need anything when time comes.

One thing I don't understand or maybe misunderstanding (nice weather induced migraine isn't helping currently).

So I have a pulse meter running on the electric meter which gives good data, so when I try using bigger time periods, or for that matter the 10 minutes each hour the cap kicks in, I never get the missing usage between the gaps.

Eg boil a kettle during a gap, that usage never appears.

So it feels like I only get the current/last usage Vs and no accumulation between the polls.

Likewise when I had a larger gap due to the gap, and I think this confirms what you've said,  often I got no gas data at all.  

So are we saying the same thing? 

Life would be easier if a) they do as they suggested that would and increase the cap and b) local data feed via mqtt server (much more long term solution I suspect).

**Reply by BottlecapDave** [reply]
> Eg boil a kettle during a gap, that usage never appears.

The data should appear in the `current_accumulative_consumption` data as that is the consumption for the current day. The `current_accumulative` data should show the delta of consumption recorded since the last successful retrieval.

As mentioned in the docs, if you set the polling configuration to 2 minutes, this should be enough to poll data for both electricity and gas without hitting the rate limits (with a bit of a buffer for any app activity which also counts against rate limits). This is assuming you only have a single instance of the integration running.

Gas data will come in bigger batches as it doesn't seem to be captured as granularly. I think this is a limitation in how the smart meters capture the data.

If you believe data is missing, the best thing to confirm is if the data is the same in the app. If you look at "today", you'll get the data from the Mini. 
* If the total matches the value from `current_accumulative_consumption` (give or take a bit to account for polling), then the data is being captured. 
* If the data in the app has a big difference, then there might be another issue with the integration for which I would need debug logs to try and track down the issue.
* If there is data missing in both the app and the integration, then that would suggest the Home Mini has failed to publish the data

**Reply by Jilas** [reply]
Thanks the might well explain what I was missing, with regards to the data, I'll check tomorrow.

The gas meters by design only send data every 30 minutes or so to save on battery power - now if only they could say when in the half hour it was.  

Thanks as always.

**Comment by andyblac** [comment]
@BottlecapDave thanks for the electric and gas update separation, I have set my GAS to 5 mins see what happens.

**Reply by Jilas** [reply]
@BottlecapDave I second that thanks. 

Using the gas check at 2 minutes and already recovered the missing 10 minutes.   Gas will take longer to check, but fingers crossed.


I've chased octopus but am currently getting static back from them which is a shame.  I'll keep trying on the off chance that they are listening.

**Reply by BottlecapDave** [reply]
You might be getting radio silence as it _sounds_ like they're working on MQTT support https://x.com/agile_phil/status/1702320398157885851?s=20

**Reply by Jilas** [reply]
> You might be getting radio silence as it _sounds_ like they're working on MQTT support https://x.com/agile_phil/status/1702320398157885851?s=20

Interesting we can hope but was told the change was requested, and regards mqtt and the party line was never say never to mqtt support.

The disappointment is someone that high up reaching out, offering a solution (that hasn't occurred so far), and asking for additional feedback, to then get silence is a let down.  Much prefer regular updates (even with no info, but at least you can trust they are monitoring), and then under promise and over deliver.

They may do the latter but without the updates they have lost the confidence.

Let's see what time brings I guess.

But for now, your splitting of the timers appears to be working.  Thanks.
